pygbm/loss.py

```diff
@@ -104,5 +114,82 @@ def _update_gradients_hessians_logistic(gradients, hessians, y_true, y_pred):
+
+    def __call__(self, y_true, y_pred, average=True):
+        y_pred = y_pred.reshape(self.n_trees_per_iteration, -1).T
+        one_hot_true = np.zeros_like(y_pred)
+        for k in range(self.n_trees_per_iteration):
+            one_hot_true[:, k] = (y_true == k)
+
+        return (logsumexp(y_pred, axis=1) -
+                (one_hot_true * y_pred).sum(axis=1))
+
+    def update_gradients_and_hessians(self, gradients, hessians, y_true,
+                                      y_pred, k=0):
+        # reshape will hopefully return a view so it's cheap?
+        # reshape y_pred to (n_samples, n_trees_per_iteration)
+        y_pred = y_pred.reshape(self.n_trees_per_iteration, -1).T
+        return _update_gradients_hessians_softmax(gradients, hessians,
+                                                  y_true, y_pred,
+                                                  self.n_trees_per_iteration)
+
+    def predict_proba(self, decision):
+        # TODO: This could be done in parallel
+        decision = decision.reshape(self.n_trees_per_iteration, -1).T
+        # compute softmax (using exp(log(softmax)))
+        proba = np.exp(decision - logsumexp(decision, axis=1)[:, np.newaxis])
+        return proba
+
+
+@njit(parallel=True)
+def _update_gradients_hessians_softmax(gradients, hessians, y_true, y_pred,
+                                       n_trees_per_iteration):
+    # gradients, hessians, and y_pred are of shape
+    # (n_samples * n_trees_per_iteration,).
+    # y_true is of shape (n_samples,).
+    #
+    # instead of passing the whole gradients and hessians arrays and slicing
+    # them here, we could instead do the update in the 'for k in ...' loop of
+    # fit(), by passing gradients_at_k and hessians_at_k which are of size
+    # (n_samples,).
+    # That would however require to pass a copy of y_pred, so it does not get
+    # partially overwritten at the end of the loop when _update_y_pred() is
+    # called.
+    n_samples = y_true.shape[0]
+    starts, ends, n_threads = _get_threads_chunks(total_size=n_samples)
+    for k in range(n_trees_per_iteration):
+        gradients_at_k = gradients[n_samples * k:n_samples * (k + 1)]
+        hessians_at_k = hessians[n_samples * k:n_samples * (k + 1)]
+        for thread_idx in prange(n_threads):
+            for i in range(starts[thread_idx], ends[thread_idx]):
+                # p_k is the probability that class(ith sample) == k.
+                # This is a regular softmax. Ideally we should be using
+                # logaddexp or logsumexp but they are not supported by numba.
+                denominator = np.sum(np.exp(y_pred[i, :]))
+                p_k = np.exp(y_pred[i, k] - np.log(denominator))
+                gradients_at_k[i] = p_k - (y_true[i] == k)
+                hessians_at_k[i] = p_k * (1. - p_k)
+                # LightGBM uses 2 * p_k * (1 - p_k) but that seems incorrect?
+                # Looks like it gives much better results in practice though.
```

Comment:
@ogrisel This is ready for reviews, but this discrepancy in particular is confusing to me

Reply:
Have you tried to compare with different values of the learning rate? Doubling the hessian is equivalent to decreasing the learning rate by 2, isn't it?

Reply:
You are right, dividing the learning rate by 2 should gives the same results (as well as `min_hessian_to_split` but that's not critical here). And it does give the same results, provided that we change 

`shrinkage = 1. if self.n_iter_ == 0 else self.learning_rate`

to just

`shrinkage = self.learning_rate`.

Setting the shrinkage it to 1 for the first iteration seems to hurt accuracy quite a bit, and [LightGBM](https://github.com/Microsoft/LightGBM/blob/master/src/boosting/gbdt.cpp#L371) applies shrinkage at every iteration including the first one.

Should we change this?

Reply:
Yes. I did that because I thought that LightGBM was doing but maybe I misread the LightGBM code.